### PR TITLE
coq_makefile: make sure compile flags for Coq and coq_makefile are in sync (supersed #1039)…

### PR DESCRIPTION
**Pull request by gares**
E.g. -safe-string is set by configure.ml and made available to
both make (via config/Makefile) and coq_makefile (via
config/coq_config.ml -> lib/envars.ml -> CoqMakefile.in).

### CODE REVIEW

**Comment by gares** [comment]
TODO:
- [ ] coqmktop still hardcodes the flags
- [ ] it compiled here, but I think I've lost `-thread` in the process, so something is still fishy

**Comment by ejgallego** [comment]
If you are gonna go this route IMHO we should take advice from @gasche in #540 .

File: configure.ml [hunk 1]
```diff
@@ -1156,7 +1156,7 @@ let write_makefile f =
   pr "CAMLHLIB=%S\n\n" camllib;
   pr "# Caml link command and Caml make top command\n";
   pr "# Caml flags\n";
-  pr "CAMLFLAGS=-rectypes %s %s %s\n" coq_warn_flags coq_annotate_flag coq_safe_string;
+  pr "CAMLFLAGS=%s\n" caml_flags;
```
Comment (ejgallego) [inline configure.ml]: Umm, why is this stuff printed twice by configure and `envvars.ml` ?

File: tools/CoqMakefile.in [hunk 2]
```diff
@@ -39,6 +39,7 @@ CAMLP4O           := $(COQMF_CAMLP4O)
 CAMLP4BIN         := $(COQMF_CAMLP4BIN)
 CAMLP4LIB         := $(COQMF_CAMLP4LIB)
 CAMLP4OPTIONS     := $(COQMF_CAMLP4OPTIONS)
+CAMLFLAGS         := $(COQMF_CAMLFLAGS)
```
Comment (ejgallego) [inline tools/CoqMakefile.in]: What happens if the user overrides this variable?

File: tools/coqmktop.ml [hunk 3]
```diff
@@ -272,7 +272,7 @@ let main () =
     (* - We add topstart.cmo explicitly because we shunted ocamlmktop wrapper.
        - With the coq .cma, we MUST use the -linkall option. *)
     let args =
-      "-linkall" :: "-rectypes" :: "-w" :: "-31" :: flags @ copts @ options @
+      "-linkall" :: "-rectypes" :: "-safe-string" :: "-w" :: "-31" :: flags @ copts @ options @
```
Comment (ejgallego) [inline tools/coqmktop.ml]: Indeed it seems that we could just take this from `Coq_config` in the style of #540

**Comment by SkySkimmer** [comment]
Not sure about putting the `-warn-flags` in coq_makefile. If I compile Coq with warn-error that doesn't imply I want to compile plugins with it.

**Comment by ejgallego** [comment]
Good point @SkySkimmer , certainly warnings should go into their own variable.

File: tools/CoqMakefile.in (hunk 2 above)
Comment (gares) [reply tools/CoqMakefile.in]: You are welcome to be more explicit here:
  https://github.com/coq/coq/blob/master/doc/refman/RefMan-uti.tex#L132
and suggest to use `CAMLFLAGS+=extra`

File: configure.ml (hunk 1 above)
Comment (gares) [reply configure.ml]: Not sure I understand the question.

FTR:
./configure generates config/coq_config.ml and config/Makefile.
envars is a wrapper around coq_config.ml letting the user override some values by setting appropriate environment variables.

**Comment by gares** [comment]
@SkySkimmer please check my last commit works for you

File: tools/coqmktop.ml [hunk 4]
```diff
@@ -271,8 +271,10 @@ let main () =
   try
     (* - We add topstart.cmo explicitly because we shunted ocamlmktop wrapper.
        - With the coq .cma, we MUST use the -linkall option. *)
+    let coq_camlflags =
+      List.filter ((<>) "") (String.split_on_char ' ' Coq_config.caml_flags) in
     let args =
-      "-linkall" :: "-rectypes" :: "-w" :: "-31" :: flags @ copts @ options @
+      coq_camlflags @ "-linkall" :: "-w" :: "-31" :: flags @ copts @ options @
```
Comment (gares) [inline tools/coqmktop.ml]: @maximedenes can you check this -w-31 is OK?

**Comment by SkySkimmer** [comment]
LGTM

File: tools/CoqMakefile.in (hunk 2 above)
Comment (ejgallego) [reply tools/CoqMakefile.in]: So not possible to do `make CAMLFLAGS=foo` ?

**Comment by gares** [comment]
> If you are gonna go this route IMHO we should take advice from @gasche in #540 .

I tried to read this PR but I'm a bit lost, what do you suggest concretely? Is it really about this PR or worth another refinement iteration?

File: tools/CoqMakefile.in (hunk 2 above)
Comment (gares) [reply tools/CoqMakefile.in]: Hum, I think that "we" know better the flags to pass to `ocamlc`, e.g. we know that we must pass `-safe-string` or things will break (that is the point of this PR). IMHO coq_makefile should work for the lambda user, and I quite don't see why they would like to override completely the flags.  I mean, adding a -I or something like that is still possible (and somewhat documented).

More in general, I did not document nor think too much about which flags can be overridden on the command line. The only few that are documented/thought-about are the timing ones and VERBOSE.  I guess we should say explicitly that all variables not documented as "overrideable" are not supposed to be overridden (in the doc/makefile header).

**Comment by Zimmi48** [comment]
@gares Can you fix your PR? It doesn't compile with OCaml 4.02.3.

**Comment by ejgallego** [comment]
> I tried to read this PR but I'm a bit lost, what do you suggest concretely? Is it really about this PR or worth another refinement iteration?

Sorry for the late reply Enrico, I was indeed suggesting we follow [this](https://github.com/coq/coq/pull/540#issuecomment-292180001) concrete piece of information wrt flag organization.

**Pull request merged**